File: pkgs/vm-runner/src/virby_vm_runner/circuit_breaker.py
"""Circuit breaker pattern for VM operations."""

import time
from enum import Enum
from typing import Any, Callable

from .exceptions import VMRuntimeError


class CircuitState(Enum):
    """Circuit breaker states."""

    CLOSED = "closed"  # Normal operation
    OPEN = "open"  # Failing, reject requests
    HALF_OPEN = "half_open"  # Testing if service recovered


class CircuitBreaker:
    """Circuit breaker to prevent cascading failures in VM operations."""

    def __init__(self, failure_threshold: int = 5, timeout: float = 30.0):
        """Initialize circuit breaker.

        Args:
            failure_threshold: Number of failures before opening circuit
            timeout: Time to wait before attempting recovery (seconds)
        """
        self.failure_threshold = failure_threshold
        self.timeout = timeout
        self.failure_count = 0
        self.last_failure_time = None
        self.state = CircuitState.CLOSED

    async def call(self, func: Callable, *args, **kwargs) -> Any:
        """Execute function via circuit breaker.

        Args:
            func: Function to execute
            *args: Function arguments
            **kwargs: Function keyword arguments

        Returns:
            Function result

        Raises:
            VMRuntimeError: If circuit is open or function fails
        """
        if self.state == CircuitState.OPEN:
            if time.time() - self.last_failure_time > self.timeout:
                self.state = CircuitState.HALF_OPEN
            else:
                raise VMRuntimeError("Circuit breaker is OPEN")

        try:
            result = await func(*args, **kwargs)
            self._on_success()
            return result
        except Exception:
            self._on_failure()
            raise

    def _on_success(self):
        """Handle successful operation."""
        self.failure_count = 0
        self.state = CircuitState.CLOSED

    def _on_failure(self):
        """Handle failed operation."""
        self.failure_count += 1
        self.last_failure_time = time.time()

        if self.failure_count >= self.failure_threshold:
            self.state = CircuitState.OPEN

    def reset(self):
        """Manually reset circuit breaker to closed state."""
        self.failure_count = 0
        self.last_failure_time = None
        self.state = CircuitState.CLOSED

    @property
    def is_open(self) -> bool:
        """Check if circuit is open."""
        return self.state == CircuitState.OPEN

    @property
    def is_half_open(self) -> bool:
        """Check if circuit is half-open."""
        return self.state == CircuitState.HALF_OPEN

    @property
    def is_closed(self) -> bool:
        """Check if circuit is closed."""
        return self.state == CircuitState.CLOSED

File: pkgs/vm-runner/internal/circuit_breaker/circuit_breaker.go
package circuit_breaker

import (
	"fmt"
	"sync"
	"time"
)

type CircuitState int

const (
	CircuitStateClosed CircuitState = iota
	CircuitStateHalfOpen
	CircuitStateOpen
)

type CircuitBreaker struct {
	failureCount     int
	failureThreshold int
	lastFailureTime  time.Time
	mu               sync.Mutex
	state            CircuitState
	timeout          time.Duration
}

func NewCircuitBreaker(failureThreshold int, timeout time.Duration) *CircuitBreaker {
	return &CircuitBreaker{
		failureThreshold: failureThreshold,
		timeout:          timeout,
	}
}

func (c *CircuitBreaker) Call(fn func() error) error {
	c.mu.Lock()
	defer c.mu.Unlock()
	if c.state == CircuitStateOpen {
		if time.Since(c.lastFailureTime) > c.timeout {
			c.state = CircuitStateHalfOpen
		} else {
			return fmt.Errorf("circuit breaker is open")
		}
	}

	if err := fn(); err != nil {
		c.lastFailureTime = time.Now()
		c.failureCount++
		if c.failureCount >= c.failureThreshold {
			c.state = CircuitStateOpen
		}
		return err
	}

	c.failureCount = 0
	c.state = CircuitStateClosed
	return nil
}

func (c *CircuitBreaker) IsClosed() bool {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.state == CircuitStateClosed
}

func (c *CircuitBreaker) IsHalfOpen() bool {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.state == CircuitStateHalfOpen
}

func (c *CircuitBreaker) IsOpen() bool {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.state == CircuitStateOpen
}

func (c *CircuitBreaker) Reset() {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.failureCount = 0
	c.lastFailureTime = time.Time{}
	c.state = CircuitStateClosed
}
